### PR TITLE
Ported Huggingface LambdaLR-based schedulers

### LINKED ISSUE #4044: Port learning rate schedulers from Huggingface
Huggingface contains a bunch of custom ```LambdaLR``` based LR schedulers they use to get scores for different tasks.
We could simply port them so they can be instantiated from the configuration file. 

Here is a working example of how to register one of the schedulers:
```python

from typing import Any, Dict, List, Union

from allennlp.common.registrable import Registrable
from allennlp.training.optimizers import Optimizer
from allennlp.training.learning_rate_schedulers import LearningRateScheduler
from allennlp.training.learning_rate_schedulers.learning_rate_scheduler import  _PyTorchLearningRateSchedulerWrapper

from transformers import get_linear_schedule_with_warmup

@LearningRateScheduler.register("linear_with_warmup")
class WarmpuLinearLearningRateScheduler(_PyTorchLearningRateSchedulerWrapper):
    """
    Registered as a `LearningRateScheduler` with name "linear_with_warmup".
    """

    def __init__(
        self, optimizer: Optimizer, num_warmup_steps: int, num_training_steps: int, last_epoch: int = -1
    ) -> None:
        lr_scheduler = get_linear_schedule_with_warmup(
            optimizer=optimizer, 
            num_warmup_steps=num_warmup_steps, 
            num_training_steps=num_training_steps, 
            last_epoch=last_epoch)
        super().__init__(lr_scheduler)
```
If you like the idea, do you think we should add these at the end of the ```allennlp.training.learning_rate_schedulers``` file, or create a separate file (e.g. ``` huggingface_learning_rate_schedulers```)?

The function  ```get_linear_schedule_with_warmup``` returns regular pytroch ```lr scheduler```.

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Fixes #4044 .

Changes proposed in this pull request:

- Ported the following Huggingface `LambdaLR`-based schedulers: `ConstantLearningRateScheduler`, `ConstantWithWarmupLearningRateScheduler`, `CosineWithWarmupLearningRateScheduler`, `CosineHardRestartsWithWarmupLearningRateScheduler`

### CODE REVIEW

File: tests/training/learning_rate_schedulers/learning_rate_scheduler_test.py [hunk 1]
```diff
@@ -138,3 +138,48 @@ def test_exponential_works_properly(self):
         assert optimizer.param_groups[0]["lr"] == 0.5 ** 2
         scheduler.step()
         assert optimizer.param_groups[0]["lr"] == 0.5 ** 3
+
+    def test_huggingface_schedulers_work_properly(self):
```
Comment (epwalsh) [inline tests/training/learning_rate_schedulers/learning_rate_scheduler_test.py:142]: 💯

**Review by epwalsh — APPROVED** [review]
Looks great! Good job with the tests!

File: allennlp/training/learning_rate_schedulers/learning_rate_scheduler.py [hunk 2]
```diff
@@ -129,3 +136,79 @@ def __init__(
...
+@LearningRateScheduler.register("constant")
+class ConstantLearningRateScheduler(_PyTorchLearningRateSchedulerWrapper):
+    """
+    Registered as a `LearningRateScheduler` with name "constant".  The "optimizer" argument does
+    not get an entry in a configuration file for the object.
+    """
+
+    def __init__(self, optimizer: Optimizer, last_epoch: int = -1) -> None:
+        lr_scheduler = get_constant_schedule(optimizer=optimizer, last_epoch=last_epoch)
+        super().__init__(lr_scheduler)
```
Comment (dirkgr) [inline allennlp/training/learning_rate_schedulers/learning_rate_scheduler.py:150]: It's fine to have this for completeness, but I'm pretty sure you could already get a constant learning rate if you just don't specify anything.

**Review by dirkgr — APPROVED** [review]

**Comment by dirkgr** [comment]
I see you're working from your own fork. That's fine if you want to work that way, but you can just create branches on the main repo if you like.
